### DISCUSSION TITLE
Restrict Mojo's parameters only to fn / struct

### DISCUSSION THREAD

**Discussion by gryznar** [modular/modular#498] [Ideas]
As explained by Nick here: https://discuss.python.org/t/pep-695-type-parameter-syntax/21646/125 Mojo's parameters are not compatible with Python syntax to declare generic subtypes.

```python
def foo[T: int](bar: T, baz: T): ... 
```
In Python code above means, that you should pass 2 `int` types when calling foo, eg.: `foo(1, 2)`

```mojo
fn foo[T: Int](): ...
```
In Mojo code above means, that you should pass parameter which is constant at compile time, e.g. `foo[1]()`

As you can see Mojo is extending Python in incompatible way. Maybe a better solution in this case will be to restrict Mojo parameters only to `fn` | `struct` and treat it in `def` and in future `class` like Python does?

Extending only Mojo structures and keeping `def` without changes comparing to Python will give also compatibility with future changes in Python.

**Comment by lattner** [comment]
Yep, this is a known issue.  We can address this in a few different ways, the easiest of which is to do name lookup on 'int' and see that it isn't defined as a `struct` or `class` and then handle it differently.

**Reply by gryznar** [reply]
So in this case if I am not wrong `def` and `class` won't be parametrized as is in Python? :)

**Reply by lattner** [reply]
No, I think it is important that classes (and def's, though that is negotiable for sure) are parameterizable.  I'm suggesting that in:

```
from typing import Dict
```

we know that `Dict` is a random variable, not a declared Mojo type.  Same holds for `int`.

**Reply by gryznar** [reply]
So parametrization will be or won't be according to type? If type is Mojo's struct / primitive, parametrization will take place, in other case won't?


I am not against that, but it can be strange and confusing for people / tools that is some cases it is required parameter, while in other - annotation. Some libraries may also relate on these CPython annotations.


Extending `def` and `class` can also have some real impact on future development and compatibility with Python. 

Let's consider some Mojo's only related feature (related to compilation for example). Introducing it to `fn` / `struct` is painless and straightforward. On the other hand introducing it to `class` / `def` could introduce danger, that CPython will use similar implementation to other things and this will break compatibility.


To summarize, extending Mojo's stuff is safe and maintainable, while extending Python's could have lamentable consequences. Keeping that in sync could also prevent from adding some cool stuff to Mojo.

**Comment by lattner** [comment]
I'm not suggesting a specific solution here, I'm agreeing this is an issue and saying that there are multiple ways to solve it.

**Reply by gryznar** [reply]
Okay :) Thank you for clarification!

**Comment by s3drik** [comment]
Maybe using decorators will help.

**Reply by gryznar** [reply]
Personally I like current syntax, it is much cleaner
